### PR TITLE
backend/enh: stripe ineg in ride flow

### PR DESCRIPTION
**Pull request by chakri642**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Screenshot of test results
Screenshot 2024-08-20 at 2 29 02 PM

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added integration tests for my changes where possible
- [ ] I tested the changes using integration tests
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

**Review by DwivediAman-exe — DISMISSED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/RidePayment.hs [hunk 1]
```diff
@@ -161,35 +161,15 @@ postPaymentAddTip ::
...
+postPaymentAddTip (_mbPersonId, merchantId) rideId tipRequest = do
   Redis.whenWithLockRedis addTipLockKey 60 $ do
-    personId <- mbPersonId & fromMaybeM (PersonNotFound "No person found")
-    person <- runInReplica $ QPerson.findById personId >>= fromMaybeM (PersonNotFound personId.getId)
     ride <- runInReplica $ QRide.findById rideId >>= fromMaybeM (RideNotFound rideId.getId)
     unless (ride.status == Domain.Types.Ride.COMPLETED) $
       throwError $ RideInvalidStatus ("Ride is not completed yet." <> Text.pack (show ride.status))
     fareBreakups <- runInReplica $ QFareBreakup.findAllByEntityIdAndEntityType rideId.getId Domain.Types.FareBreakup.RIDE
     when (any (\fb -> fb.description == tipFareBreakupTitle) fareBreakups) $ throwError $ InvalidRequest "Tip already added"
-    customerPaymentId <- person.customerPaymentId & fromMaybeM (PersonFieldNotPresent "customerPaymentId")
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/RidePayment.hs:173]: if the tip is added after the scheduler run, then tip should be charged separately

File: Backend/lib/payment/src/Lib/Payment/Domain/Action.hs [hunk 2]
```diff
@@ -128,7 +129,7 @@ createPaymentIntentService ::
   m Payment.CreatePaymentIntentResp
 createPaymentIntentService merchantId personId rideId rideShortIdText createPaymentIntentReq createPaymentIntentCall updatePaymentIntentAmountCall capturePaymentIntentCall getPaymentIntentCall = do
   let rideShortId = ShortId rideShortIdText
-  let cardFixedCharges = HighPrecMoney 0.3 -- 30 cents, we have to fix it later properly
+  -- let cardFixedCharges = HighPrecMoney 0.3 -- 30 cents, we have to fix it later properly
```
Comment (NikithShetty) [inline Backend/lib/payment/src/Lib/Payment/Domain/Action.hs:132]: Remove commented out code

**Review by NikithShetty — CHANGES_REQUESTED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/RidePayment.hs [hunk 3]
```diff
@@ -162,31 +162,40 @@ postPaymentAddTip ::
     Environment.Flow APISuccess
   )
 postPaymentAddTip (mbPersonId, merchantId) rideId tipRequest = do
-  Redis.whenWithLockRedis addTipLockKey 60 $ do
+  Redis.withWaitOnLockRedisWithExpiry paymentJobExecLockKey 60 60 $ do
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/RidePayment.hs:165]: Update to 10, 20 durations

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+              driverAccountId
+            }
+
+    booking <- runInReplica $ QRB.findById ride.bookingId >>= fromMaybeM (BookingDoesNotExist ride.bookingId.getId)
+    paymentIntentResp <- SPayment.makePaymentIntent person.merchantId person.merchantOperatingCityId person.id ride createPaymentIntentReq
+    void $ SPayment.chargePaymentIntent booking.merchantId booking.merchantOperatingCityId paymentIntentResp.paymentIntentId
+    QRide.markPaymentStatus DRide.Completed ride.id
+  return Complete
+  where
+    paymentJobExecLockKey :: Text
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs:84]: move this to common place

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/RidePayment.hs [hunk 5]
```diff
@@ -162,31 +162,40 @@ postPaymentAddTip ::
...
+        -- we will add this tip amount in ride end and charge it in job which is already created in ride end
+        QRide.updateTipByRideId (Just $ mkPrice (Just tipRequest.amount.currency) tipRequest.amount.amount) rideId -- update tip in ride
+      else do
+        -- Here we creating a new payment intent for tip if the ride payment status is already initiated
+        customerPaymentId <- person.customerPaymentId & fromMaybeM (PersonFieldNotPresent "customerPaymentId")
+        paymentMethodId <- person.defaultPaymentMethodId & fromMaybeM (PersonFieldNotPresent "defaultPaymentMethodId")
+        driverAccountId <- ride.driverAccountId & fromMaybeM (RideFieldNotPresent "driverAccountId")
+        email <- mapM decrypt person.email
+        let cardFixedCharges = HighPrecMoney 0.3
+        let cardPercentageCharges = 0.029 -- 2.9%
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/RidePayment.hs:184]: Make application fee computation common

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  ( EncFlow m r,
+    CacheFlow m r,
+    MonadFlow m,
+    EsqDBFlow m r,
+    SchedulerFlow r,
+    HasShortDurationRetryCfg r c
+  ) =>
+  Job 'ExecutePaymentIntent ->
+  m ExecutionResult
+sendExecutePaymentIntent Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs:49]: rename to `executePaymentIntentJob`

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+              currency = fareWithTip.currency,
+              customer = customerPaymentId,
+              paymentMethod = paymentMethodId,
+              receiptEmail = email,
+              driverAccountId
+            }
+
+    booking <- runInReplica $ QRB.findById ride.bookingId >>= fromMaybeM (BookingDoesNotExist ride.bookingId.getId)
+    paymentIntentResp <- SPayment.makePaymentIntent person.merchantId person.merchantOperatingCityId person.id ride createPaymentIntentReq
+    void $ SPayment.chargePaymentIntent booking.merchantId booking.merchantOperatingCityId paymentIntentResp.paymentIntentId
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs:80]: Update `chargePaymentIntent` to return charge status and then update the payment status accordingly

File: Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+  person <- runInReplica $ QPerson.findById personId >>= fromMaybeM (PersonNotFound personId.getId)
+  ride <- runInReplica $ QRide.findById rideId >>= fromMaybeM (RideNotFound rideId.getId)
+  merchant <- CQM.findById booking.merchantId >>= fromMaybeM (MerchantNotFound booking.merchantId.getId)
+  merchantOpCity <- CQMOC.findById booking.merchantOperatingCityId >>= fromMaybeM (MerchantOperatingCityNotFound booking.merchantOperatingCityId.getId)
+  order <- QOrder.findById (cast rideId) >>= fromMaybeM (PaymentOrderNotFound rideId.getId)
+  mobileCountryCode <- person.mobileCountryCode & fromMaybeM (PersonFieldNotPresent "mobileCountryCode")
+  mobileNumber <- mapM decrypt person.mobileNumber >>= fromMaybeM (PersonFieldNotPresent "mobileNumber")
+  when (isNothing ride.cancellationFeeIfCancelled) $ do
+    QRide.updateCancellationFeeIfCancelledField (Just cancellationAmount.amount) rideId
+  void $ SPayment.makeCancellationPayment booking.merchantId booking.merchantOperatingCityId order.paymentServiceOrderId cancellationAmount.amount
```
Comment (NikithShetty) [inline Backend/app/rider-platform/rider-app/Main/src/SharedLogic/Scheduler/Jobs/ExecutePaymentIntent.hs:115]: Need to check `makeCancellationPayment` status to update payment status

**Review by NikithShetty — CHANGES_REQUESTED** [review]

**Review by NikithShetty — APPROVED** [review]

**Pull request merged**